core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt

```diff
@@ -0,0 +1,177 @@
+      } else if (!File(storage).exists()) {
+        return defaultStorage()
+      }
+      return storage
+    }
+
+  private fun defaultStorage(): String =
+    getExternalFilesDirs(instance, null)[0]?.path ?: instance.filesDir.path
+
+  fun getPrefStorageTitle(defaultTitle: String): String =
+    sharedPreferences.getString(PREF_STORAGE_TITLE, "")?.toString() ?: defaultTitle
+
+  fun putPrefLanguage(language: String) =
+    sharedPreferences.edit { putString(PREF_LANG, language) }
+
+  fun putPrefIsFirstRun(isFirstRun: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_IS_FIRST_RUN, isFirstRun) }
+
+  fun putPrefWifiOnly(wifiOnly: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_WIFI_ONLY, wifiOnly) }
+
+  fun putPrefStorageTitle(storageTitle: String) =
+    sharedPreferences.edit { putString(PREF_STORAGE_TITLE, storageTitle) }
+
+  fun putPrefStorage(storage: String) {
+    sharedPreferences.edit { putString(PREF_STORAGE, storage) }
+    prefStorages.onNext(storage)
+  }
+
+  fun getPrefStorages(): Flowable<String> = prefStorages.startWith(prefStorage)
+
+  fun putPrefFullScreen(fullScreen: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_FULLSCREEN, fullScreen) }
+
+  fun putPrefExternalLinkPopup(externalLinkPopup: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_EXTERNAL_LINK_POPUP, externalLinkPopup) }
+
+  fun showIntro(): Boolean = sharedPreferences.getBoolean(PREF_SHOW_INTRO, true)
+
+  fun setIntroShown() = sharedPreferences.edit { putBoolean(PREF_SHOW_INTRO, false) }
+
+  var showHistoryAllBooks: Boolean
+    get() = sharedPreferences.getBoolean(PREF_SHOW_HISTORY_ALL_BOOKS, true)
+    set(prefShowHistoryAllBooks) {
+      sharedPreferences.edit { putBoolean(PREF_SHOW_HISTORY_ALL_BOOKS, prefShowHistoryAllBooks) }
+    }
+
+  var showBookmarksAllBooks: Boolean
+    get() = sharedPreferences.getBoolean(PREF_SHOW_BOOKMARKS_ALL_BOOKS, true)
+    set(prefShowBookmarksFromCurrentBook) = sharedPreferences.edit {
+      putBoolean(PREF_SHOW_BOOKMARKS_ALL_BOOKS, prefShowBookmarksFromCurrentBook)
+    }
+
+  val nightMode: NightModeConfig.Mode
+    get() = from(
+      sharedPreferences.getString(PREF_NIGHT_MODE, null)?.toInt()
+        ?: AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
+    )
+
+  fun nightModes(): Flowable<NightModeConfig.Mode> = nightModes.startWith(nightMode)
+
+  fun updateNightMode() = nightModes.offer(nightMode)
+
+  var hostedBooks: Set<String>
+    get() = sharedPreferences.getStringSet(PREF_HOSTED_BOOKS, null)?.toHashSet() ?: HashSet()
+    set(hostedBooks) {
+      sharedPreferences.edit { putStringSet(PREF_HOSTED_BOOKS, hostedBooks) }
+    }
+
+  var textZoom: Int
+    get() = sharedPreferences.getInt(TEXT_ZOOM, DEFAULT_ZOOM)
+    set(textZoom) {
+      sharedPreferences.edit { putInt(TEXT_ZOOM, textZoom) }
+      textZooms.offer(textZoom)
+    }
+
+  fun getTextZooms(): Flowable<Int> = textZooms.startWith(textZoom)
+
+  companion object {
+    // Prefs
```

Comment:
And regarding all the constants in the companionObject, i read somewhere that these are never garbage collected and hence serve as heap, so are they causing extra memory usage? I'm not pretty sure about this, i but fell into a debate with this same topic, please enlighten

Reply:
`object`s are singeltons so this companion object will exist the lifetime of our application once it is instantiated by a class load. This is standard to how `static final` fields in java work.
We don't particularly care that a couple of strings that don't take much memory at all are sticking around.
We would care if I had something like this

```
object AnAwfulObject{
var myActivity:MyActivity?=null
}

class MyActivity{
fun onCreate(){
AnAwfulObject.myActivity=this
}
```

because it holds a very leaky reference to an entire Activity which could be huge